### ISSUE TITLE
Deprecate stream ciphers with insufficient IV length

### ISSUE THREAD

**Issue by riobard** [shadowsocks/shadowsocks-org#36]
### Summary

For Shadowsocks deployment using stream ciphers, long-term key and randomly generated IV of insufficient length causes (key, IV) pair reuse with high probability, which allows Reused Key Attack. Adversaries can recover plaintext within reasonable budget constraint.

### Proposal

- Stream ciphers with IV length less than 12 bytes MUST be deprecated. Namely, `bf-cfb`, `chacha20`, and `salsa20`.
- Users SHOULD choose stream ciphers with IV length of 16 or more bytes.
- Users are RECOMMENDED to use the new AEAD ciphers.

**Comment by madeye** [comment]
I think we can show warnings, but not deprecate them.

**Comment by riobard** [comment]
I suggest we split this table https://shadowsocks.org/en/spec/cipher.html into four sections: 

1. RECOMMEND for new AEAD ciphers.
2. PHASING OUT for stream ciphers with sufficient IV length and strong security.
3. DEPRECATED for stream ciphers not in 2 or 4.
4. DANGEROUS for stream ciphers with insufficient IV length and weak security.

**Comment by riobard** [comment]
Additionally, non-IETF variant of Chacha20-Poly1305 should also be removed right now in order to prevent widespread use.

**Comment by madeye** [comment]
A pull request? I'm OK with the change of that page.

**Comment by riobard** [comment]
Working on it. 

Could you please kill chacha20-poly1305 so no one uses it? So I can remove it from that page and no need for future deprecation.

**Comment by madeye** [comment]
Fixed via https://github.com/shadowsocks/shadowsocks-libev/commit/894eae567cff1f9914d3f64223c790f93bb2d271

**Comment by wongsyrone** [comment]
要我说 libsodium 也不要用了，这个作者是个傻子，有明显漏洞的AEAD算法还放在里面让人用，明显没有经过专家审核过，而且 Google 和 Daniel J. Bernstein 教授都是傻子。

> Libsodium implements two versions of the ChaCha20-Poly1305 construction:
> The original construction can safely encrypt up to 2^64 messages with the same key, without any practical limit to the size of a message (up to 2^70 bytes).
The IETF variant is slightly slower. It can safely encrypt a pratically unlimited number of messages (2^96), but individual messages cannot exceed 64*(1^32)-64 bytes (approximatively 256 GB).
Both are interoperable with other crypto libaries, share the same security properties and are accessible via a similar API.

> In order to prevent nonce reuse, if a key is being reused, it is recommended to increment the previous nonce instead of generating a random nonce for each message.

> To prevent nonce reuse in a client-server pro tocol, either use different keys for each direction, or make sure that a bit is masked in one direction, and set in the other.

我也不知道上面的引用从哪里抄来的，也看不懂英文。

**Comment by madeye** [comment]
OK, I think @wongsyrone is right about nonce reuse of chacha20. If we generate nonce properly, we're safe with 8 bytes nonce here.

@riobard What do you think?

**Comment by riobard** [comment]
Sorry if this sounds offending, but @wongsyrone is completely wrong and misunderstands the libsodium document (which is NOT the best source of security advices by the way).

I'd suggest you read DJB's paper on [Xsalsa20] to gain more understanding.

Let me explain in a bit more details. Hopefully we could settle this argument once and for all.


## TLDR

- Chacha20, even with only 64-bit nonce, is safe if used properly with **ephemeral** keys and **unique** nonces.
- However Shadowsocks uses **long-term** key and **random** nonces, thus it's dangerous to use Chacha20 as the probability of nonce collision is significant.
- Long-term keys with duplicated nonces are vulnerable to Reused Key Attack, allowing adversaries to recover portions of plaintext, and in the context of Shadowsocks, identify the underlying protocol.
- Thus short nonces pose significant threat to the primary design goal of Shadowsocks: being indistinguishible from noises. 


## Stream ciphers 

Stream ciphers generate a key stream to XOR with plaintext to get ciphertext. Generation of key stream is a deterministic process: for a given (key, nonce) pair, the same key stream is produced.

Proper use of stream ciphers must ensure every (key, nonce) pair is unique. Usually we satisfy this requirement by generating a random, ephemeral key and using a monotonic counter value as nonce. Chacha20 uses 256-bit keys and 64-bit nonces. A randomly generated 256-bit key is extremely likely to be unique. We then use a monotonic incrementing 64-bit counter as nonce to form a unique 320-bit (key, nonce) pair, which contains 256-bit randomness. This allows 2^64 messages to be safely encrypted using the same key.


However, in Shadowsocks we use a fixed long-term key and randomly generated nonces, basicaly inverting the randomness property described above. Now we still have a 320-bit (key, nonce) pair, but only 64-bit randomness.

The probability of duplicates while generating 64-bit random nonces is extremely high (about 1/2^32,
see [Birthday Attack]).  In practice, no one should ever assume a 64-bit random number to be unique.



## AEAD ciphers

The case for nonce reuse with AEAD ciphers is more complicated and the consequences vary depending on AEAD construction. Some new AEAD ciphers are resistant to nonce reuse (so called misuse-resistant ciphers), claiming that if a (key, nonce) pair is reused, full security is still obtained (for example [GCM-SIV]).

 Shadowsocks currently supports AES-GCM and three variants of chacha20-poly1305 AEAD ciphers. AES-GCM ciphers have already been demonstrated to be vulnerable to nonce reuse with practical attacks on GCM in TLS (see [GCM-TLS]). For chacha20-poly1305, nonce reuse causes loss of confidentiality for messages with that nonce.

 The catch is that we are exhausting nonces at much higher rate using Shadowsocks AEAD ciphers. With the original stream ciphers, we use only one nonce/IV per TCP connection (or UDP packet). With AEAD ciphers, each TCP connection is broken up into records (or chunks), and each record/chunk uses two nonces (one for encrypting payload length and one for payload itself). Assuming maximum payload size of 16KB (best case), we are exhausting 128 nonces for every megabyte of traffic carried. 64-bit nonce space will not last long before duplicates occur.



---

 Cryptography is tough and security is hard. Don't be trapped by dogma. Think carefully.



[Xsalsa20]: https://cr.yp.to/snuffle/xsalsa-20081128.pdf
[Birthday Attack]: https://en.wikipedia.org/wiki/Birthday_attack
[GCM-SIV]: https://eprint.iacr.org/2015/102.pdf
[GCM-TLS]: https://eprint.iacr.org/2016/475.pdf

**Comment by madeye** [comment]
@riobard Thanks for the long explanation. So, the key point here is shadowsocks is using random nonce, leading to much larger posibility of nonce reuse?

The reason why TLS doesn't have this issue is that it exahanges keys at the beginining of a session and increases the nonce for each message. It ensures **ephemeral** keys and **unique** nonces?

**Comment by Mygod** [comment]
@madeye Correct.

@riobard Can you give an estimation about how many chunks of data is safe against nonce reuse attack for different nonce length? Also nonce reuse attack is very expensive because adversary has to record all traffic to identify nonce collisions and recover underlying traffic. We can assume that it will only be used when targeting specific server so this problem isn't very severe if the underlying doesn't need that kind of security.

**Comment by madeye** [comment]
Luckily, we still have chacha20-poly1305 and xchacha20-ietf-poly1305 here. Users can pick their favorite cipher from these two choices and won't blame us for lack of flexibilty... 👌

**Comment by riobard** [comment]
@madeye Correct. To be more precise, the problem is we are using random nonce with insufficient length AND long-term key. The choice of long-term key is deliberate to avoid handshake (exposing protocol), thus the only remedy is to use longer nonce.

@Mygod As explained above, randomly generated 64-bit number has a birthday bound at 2^(64/2). So in the best case we can only encrypt (2^32)/2 chunks. In practice the collision could happen much earlier, and it's not acceptable to rely on the estimation because the safety margin is too thin. 

> nonce reuse attack is very expensive because adversary has to record all traffic to identify nonce collisions and recover underlying traffic

This is sadly not true in our case. Adversaries need to only record a couple of packets per TCP connection to the same server/port because the first few packets contain nonce in plaintext and Shadowsocks request header in ciphertext with known structure. The cost to perform a successful Reused Key Attack on stream ciphers is actually very reasonable: ~10TB storage if my math is right.

**Comment by riobard** [comment]
@madeye You mean chacha20-ietf-poly1305 with 96-bit nonce. xchacha20-ietf-poly1305 is even better with its 192-bit nonce.

**Comment by madeye** [comment]
@riobard Yes, sorry for the typo.

**Comment by Mygod** [comment]
@riobard Sorry, I just realized there are two nonces here so let's use "IV" for the one used in stream cipher and "nonce" for the one used in authentication. The previous comment is talking about IV instead of nonce used in authentication.

About nonce reuse attack, I thought it's pretty hard to take advantage of. I didn't carefully read that paper but after skimming through it, it seems like nonce reuse attack can have a catastrophic effect because of the design flaws of the authentication algorithm. That's why we have a better algorithm using SIV.

**Comment by hellofwy** [comment]
Is TLS 1.3 is good for using as the cipher method?
TLS 1.3 now support 1-RTT or even 0-RTT handshakes.
TLS 1.3 will be a widely used protocol, so Shadowsocks's traffic can't be recognized from others.

**Comment by riobard** [comment]
@Mygod Actually the name "nonce" is better, as it reminds us that it should be used only once. The term "IV" is technically more accurate for stream ciphers, but it looks too innocent and it's very easy for people to forget that it has to be unique per key.

In this discussion we can distinguish the two: nonce for AEAD and IV for stream ciphers.

There's no evidence that adversaries are launching Reused Key Attack (note: only stream ciphers are vulnerable to RKA). However I'd like to increase the cost of doing so by choosing longer IV sizes. Hopefully it will blow off the adversary's budget constraint.

SIV hasn't been widely adopted yet. Cryptoanalysis is also lacking. It'll be years before we'll consider it. We have only AES-GCM and chacha20-poly1305 variants right now, and the former has know weakness with nonce reuse. Better to be careful.

**Comment by riobard** [comment]
@hellofwy Obfuscation is a separate concern. The pluggable transport will address that.

**Comment by Mygod** [comment]
@hellofwy TLS provides better privacy but is harder to set up and is not random noise indistinguishable. Shadowsocks's traffic still can be identified from normal TLS connections by using certificate in TLS. Also 1/0-RTT handshakes are used to resume connections and reuse ephemeral key exchanged previously if I'm not mistaken.

@riobard So could you give an estimation of connections that should be safe (like >99.7% or larger?) for different IV sizes?

Yes. By using SIP003, we are putting the responsibility of obfuscation/steganography off for plugins.

**Comment by Mygod** [comment]
Okay just found a [probability table](https://en.wikipedia.org/wiki/Birthday_problem#Probability_table). So by using 8 bytes for IV size, the chance of reusing IV is smaller than 10^-15 within only ~190 connections. For 12 bytes, the chance is smaller than 10^-15 within ~13 million connections.

> For comparison, 10^-18 to 10^-15 is the uncorrectable bit error rate of a typical hard disk.

**Comment by riobard** [comment]
@Mygod Yes, that's the quick lookup table for all the math mentioned in this issue.

Now you see why 8-byte IV provides very thin safety margin.

**Comment by Mygod** [comment]
However it's hard to tell how much security we (and/or the user) need against this kind of attack. 10^-15 may be too strict since it is very unlikely that the adversary is able to record all traffic.

**Comment by riobard** [comment]
I don't think it's an active threat yet. Nevertheless we should be more careful. Increasing IV size incurs minimal cost. The cost/benefit analysis is straightforward.

**Comment by wongsyrone** [comment]
After reading some 101 materials, I know your concerns on nonce's uniqueness and long-time pre-shared key.

https://www.crypto101.io/
https://www.lvh.io/posts/nonce-misuse-resistance-101.html

**Comment by madeye** [comment]
@wongsyrone Cool!

I think @riobard teach us a very important lesson about nonce-misuse issues. Thanks again for the info!

**Comment by riobard** [comment]
Glad to be helpful ^_^

**Comment by wongsyrone** [comment]
我感觉用英文表达不太清楚。

之前没有考虑reuse的原因是 每个TCP链接的首包需要生成IV，之后的数据统一使用这个IV，但是如果要上AEAD的话，因为消耗CSPRNG的entropy过快（在一些VM以及嵌入式设备上更明显），可能生成的时候出现重复，或者是生成的时候没重复，但是在之后incrementation的时候出现重复。遇到重复的nonce和相同的key，很可能生成相同keystream（如果我没理解错），很大程度导致加密无效泄漏明文的协议信息。

我个人建议之前stream cipher部分不要动，如果动的话，就在新的部分上做修改。
首先AEAD的设计和适用范围很大部分局限在TLS上，尤其是TLS 1.3要作为唯一支持的形式。但是我们并不是要重新设计TLS，我之前也是忽略了AEAD在TLS里面应用的一些隐含部分，比如session协商的nonce。

AEAD提高了对随机数质量的要求，由链接数级别提高到数据块级别，但是受限于SS自身，无法使用一些TLS里面惯常的避免方法。

把这里的AEAD做得太像TLS的话，我感觉不是SS的目标。

本来SIP004里面的draft只是弄出来玩的，分析到现在，我认为不适合作为翻墙协议存在，可以作为一般数据加密用途，尤其是使用AEAD作为加密部分，这里设计的时候没有考虑到诸多的隐含需求。

**Comment by Mygod** [comment]
@wongsyrone AEAD isn't designed for TLS 1.3. What other downsides do you think SIP004 have?

**Comment by wongsyrone** [comment]
重放攻击

因为现在shadowsocks libev是弄了个1024容量的缓存用来记录用过的nonce，这个缓存容量不可能无限大，只能维持在一定数额之内。如果第三方一直记录来往的加密包的话，在缓存满之后（一段时间后），一些旧值会被替代，这时候如果遇到重复nonce应该会当成正确包来处理。

但是TLS里面有另外的序列号方法可以用来防止重放。
一些在TLS里面的方法比如：https://github.com/WeMobileDev/article/blob/master/%E5%9F%BA%E4%BA%8ETLS1.3%E7%9A%84%E5%BE%AE%E4%BF%A1%E5%AE%89%E5%85%A8%E9%80%9A%E4%BF%A1%E5%8D%8F%E8%AE%AEmmtls%E4%BB%8B%E7%BB%8D.md#333-防重放

**Comment by wongsyrone** [comment]
Crypto101:

> nonce: Generally does not have to be secret or unpredictable, but does have
to be unique
> IV: Generally not required to be secret, but required to be unpredictable

之前：仅需不可预测，可通过合适的CSPRNG实现
之后：必须不能重复，相对难以达到，更何况生成方式从之前链接数级别升高到数据块级别，每个块需要increment两次分别加密长度信息和数据。

如果同样是1024的nonce cache，之前需要1024个链接，现在只需1024个块，极大提前了cache满的时刻。

**Comment by wongsyrone** [comment]
因为之前限定数据块长度0x3FFF，这样块数比不限制肯定是增加了，一定程度减小了重传的影响，但是加速entropy消耗，进一步减小nonce cache的保护作用。

**Comment by Mygod** [comment]
Hmm... Replay attack is also possible before SIP004 so maybe we should open a separate issue to discuss it?

**Comment by wongsyrone** [comment]
我感觉SIP004里面的陷阱多（预备知识，不论是加密学还是TLS细节），实现复杂，提升不明显，虽然有些部分比原版要所谓安全，但是我认为代价太大。

**Comment by madeye** [comment]
An easy way to defend replay is using a timestamp as the AD of our AEAD, e.g. `time() / 60 +/- 1`. As we still have two reserved bits in the 16-bit length, one bit can be used as a flag to disable/enable this timestamp.

But I don't think it's really necessary. SIP004 aims to provide IND-CCA2 and replace OTA. At least, cheap attacks by modifying our request header won't work anymore. If we decide to deprecate OTA, then SIP004 is necessary to us.

**Comment by wongsyrone** [comment]
另外一个可能是数据包的长度规律，不过没有细看过。有人有兴趣可以研究分析一下。

**Comment by riobard** [comment]
Protection from replay attack is not the job of shadowsocks. 

Statistical analysis of packet property can be obfuscated by pluggable transport.

**Comment by shinku721** [comment]
用AES时候的那个IV也是不能重复的，说到底是block cipher的时候叫IV，加上operation mode后第一个IV就是nonce了
@riobard IMO shadowsocks should get rid of request replay
@madeye the timestamp is a reasonable idea with caching of nonces in the allowed time period

**Comment by wongsyrone** [comment]
> An implementation can use any method to vary the value of the local
   64-bit integer, providing the chosen method never generates a
   duplicate IV for the same key.

>  A duplicated IV can result in the very unlikely event that multiple
   managers, communicating with a single authoritative engine, both
   accidentally select the same 64-bit integer within a second.  The
   probability of such an event is very low, and does not significantly
   affect the robustness of the mechanisms proposed.

https://tools.ietf.org/html/rfc3826#section-3.1.2.1

**Comment by madeye** [comment]
@shinku721 Yes, replay attack is still a open problem for us. Any suggestion?

**Comment by hellofwy** [comment]
@madeye 
OTA's HMAC is calculated with a Chunk ID which is an unsigned integer counted per connection from 0. I think the Chunk ID may be useful for preventing some kinds of replay attacks.
And I think a server side request id cache may be be useful for preventing request replay attacks.

**Comment by shinku721** [comment]
@madeye 
1. Just put a timestamp into the request header, then the server verifies that the timestamp should not exceed the server's local time +/- 60s. This timestamp is authenticated by the request's tag. (It is OK to use the timestamp as AD but then we cannot tell timestamp errors from authentication errors...)
2. The server keeps recording valid incoming requests (by its IV or the seed in the AEAD protocol or anything that can identify a request) for more than the last 2 minutes (2 * 60s), and rejects duplicated requests. We can use any reasonable data structures such as binary trees, hash tables or bloom filters.

Drawback: sometimes the times on clients and servers are not synchronized...

@hellofwy We are focusing on replaying of the requests... And chunks must be replayed after the attacker replays the request.

**Comment by qinyuhang** [comment]
Do u mean a UTC timestamp？bad network will have affect on it？orGFW can just
fake the one timestamp？
二階堂 真紅 <notifications@github.com>于2017年2月11日 周六15:46写道：

**Comment by hellofwy** [comment]
The time stamp part is encrypted and authenticated as other parts in AEAD ciphers, so can't be faked.

**Comment by qinyuhang** [comment]
I mean the verification ofof thetimestamp is too wide range，GFW pick 1s
probably hit a 2 min range. This cannot work on a tream communication
hellofwy <notifications@github.com>于2017年2月11日 周六17:49写道：

**Comment by shinku721** [comment]
@qinyuhang GFW requests either a manipulated header or an original header.
- A manipulated header is not authenticated.
- An original header contains the unmodified timestamp.
  * If the timestamp is out of the allowed range (GFW replays the request after 2mins), it is rejected.
  * If the timestamp is in the range, since we have cached all valid requests in the range of time, it is found to be a duplicate. So we reject it, too.

**Comment by riobard** [comment]
This thread is getting off-topic.

I've opened a new issue to discuss replay attack https://github.com/shadowsocks/shadowsocks-org/issues/44

**Comment by riobard** [comment]
Closing now that we have cipher guidance.
